plugins/debuginfo-install.py

```diff
@@ -74,59 +75,125 @@ def configure(self):
+        debugsource_suffix_len = len(Package.DEBUGSOURCE_SUFFIX)
+
         for pkgspec in self.opts.package:
-            package_query = sorted(dnf.subject.Subject(pkgspec).get_best_query(
-                self.cli.base.sack).filter(arch__neq='src'), reverse=True)
-            if not package_query:
-                msg = _('No match for argument: %s')
-                logger.info(msg, self.base.output.term.bold(pkgspec))
+            solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack)
+
+            query = solution["query"].filterm(arch__neq='src')
+
+            package_dict = query.available()._name_dict()
+            # installed versions of packages have priority, replace / add them to the dict
+            package_dict.update(query.installed()._name_dict())
+
+            if not package_dict:
+                logger.info(_('No match for argument: %s'), self.base.output.term.bold(pkgspec))
                 errors_spec.append(pkgspec)
-            for pkg in package_query:
-                self._di_install(pkg)
+                continue
 
-        if not self.dbgdone:
-            logger.info(_("No debuginfo packages available to install"))
+            # Remove debuginfo packages if their base packages are in the query.
+            # They can get there through globs and they break the installation
+            # of debug packages with the same version as the installed base
+            # packages. If the base package of a debuginfo package is not in
+            # the query, the user specified a debug package on the command
+            # line. We don't want to ignore those, so we will install them.
+            # But, in this case the version will not be matched to the
+            # installed version of the base package, as that would require
+            # another query and is further complicated if the user specifies a
+            # version themselves etc.
+            for name in list(package_dict.keys()):
+                if name.endswith(Package.DEBUGINFO_SUFFIX):
+                    if name[:-debuginfo_suffix_len] in package_dict:
+                        package_dict.pop(name)
+                if name.endswith(Package.DEBUGSOURCE_SUFFIX):
+                    if name[:-debugsource_suffix_len] in package_dict:
+                        package_dict.pop(name)
+
+            # attempt to install debuginfo and debugsource for the highest
+            # listed version of the package (in case the package is installed,
+            # only the installed version is listed)
+            for packages in package_dict.values():
+                package = packages[0]
+
+                if packages[0]._from_system:
+                    # for packages from system (installed) there can be more
+                    # packages with different architectures listed and we want
+                    # to install debuginfo for all of them
+                    for package in packages:
```

Comment:
I would recommend a different approach. It is little bit complicated therefore I provide a diff
```
diff --git a/plugins/debuginfo-install.py b/plugins/debuginfo-install.py
index 4b53565..28739d1 100644
--- a/plugins/debuginfo-install.py
+++ b/plugins/debuginfo-install.py
@@ -116,14 +116,18 @@ class DebuginfoInstallCommand(dnf.cli.Command):
             # listed version of the package (in case the package is installed,
             # only the installed version is listed)
             for packages in package_dict.values():
+                arch_dict = {}
+
                 package = packages[0]
 
-                if packages[0]._from_system:
+                if package._from_system:
                     # for packages from system (installed) there can be more
                     # packages with different architectures listed and we want
                     # to install debuginfo for all of them
-                    for package in packages:
-                        if not self._install_debug_from_system(package, package.debug_name):
+                    for pkg in packages:
+                        arch_dict.setdefault(pkg.arch, []).append(pkg)
+                    for pkgs_per_arch in arch_dict.values():
+                        if not self._install_debug_from_system(pkgs_per_arch, package.debug_name):
                             self.debuginfo_missing.add(str(package))
 
                         if not self._install_debug_from_system(package, package.debugsource_name):
@@ -131,7 +135,6 @@ class DebuginfoInstallCommand(dnf.cli.Command):
 
                 else:
                     # if the first package is not from the system, we just want its name
-                    package = packages[0]
                     if package.name.endswith(Package.DEBUGINFO_SUFFIX) \
                             or package.name.endswith(Package.DEBUGSOURCE_SUFFIX):
 
@@ -159,12 +162,14 @@ class DebuginfoInstallCommand(dnf.cli.Command):
             raise dnf.exceptions.PackagesNotAvailableError(_("Unable to find a match"),
                                                            pkg_spec=' '.join(errors_spec))
 
-    def _install_debug_from_system(self, package, debug_name):
-        query = self.base.sack.query().filter(name=debug_name,
-                                              epoch=package.epoch,
-                                              version=package.version,
-                                              release=package.release,
-                                              arch=package.arch)
+    def _install_debug_from_system(self, per_arch_packages, debug_name):
+        base_query = self.base.sack.query().filterm(arch=per_arch_packages[0].arch).apply()
+        query = self.base.sack.query().filterm(empty=True)
+        for package in per_arch_packages:
+            query.union(self.base.sack.query().filter(name=debug_name,
+                                                      epoch=package.epoch,
+                                                      version=package.version,
+                                                      release=package.release))
 
         if query:
             self._install(query)
(END)

```

Reply:
Also part
```
        base_query = self.base.sack.query().filterm(arch=per_arch_packages[0].arch).apply()
        query = self.base.sack.query().filterm(empty=True)
        for package in per_arch_packages:
            query.union(base_query.filter(name=debug_name,
                                          epoch=package.epoch,
                                          version=package.version,
                                          release=package.release))
```
could be replaced by:
```
nevras=[]
        for pkg in per_arch_packages:
            nevras.append("{}-{}:{}-{}.{}".format(
                debug_name, pkg.epoch, pkg.version, pkg.release, pkg.arch))
        query = self.base.sack.query().filterm(nevra_strict=nevras)
```
The second option would be more efficient.

Reply:
@j-mracek first, I'd really like to stress that I _need_ some sort of description of your proposed solutions. Most of the time I have to seriously try hard to figure out what you actually mean...

It seems you've got some sort of error in your patch, the `base_query` you create in `_install_debug_from_system` is never used. Even with that fixed though, I don't understand how your solution practically differs from what my code is doing...

And, on top of that, this doesn't address neither of the two problems I've described above :slightly_smiling_face::


> <snip> But the `_name_dict()` still has a problem, because the package lists in its values can contain packages with different achitectures and sometimes the wrong architecture is first in the list, and it can:
> 
> 1. be logged in case `-debuginfo` wasn't found, confusing the user
> 
> 2. in case NEVRA is None (as @j-mracek notes above), the only information we have is this package, and it is wrong

Please do see the tests in the CI repo, they demonstrate well what doesn't work. The two tests in question are marked with @xfail: [rpm-software-management/ci-dnf-stack#566](https://github.com/rpm-software-management/ci-dnf-stack/pull/566)

Reply:
So the problem actually was with installonly packages. I've added a test to the tests PR and I've fixed the code to install only the latest version for each architecture, based on the above suggestion, but I've simplified it a bit, installing only the latest package for each architecture.